### PR TITLE
Do not initialize gpg at the module level.

### LINKED ISSUE #640: pulpcore worker does not start after upgrade to 2.11
**Version**
Please provide the versions of the pulpcore and pulp_container packages in use, and how they are installed.  If you are using Pulp via Katello, please provide the Katello version.

**Describe the bug**
```
Starting gunicorn 20.1.0
Listening at: http://127.0.0.1:9000/ (522413)
Using worker: sync
Booting worker with pid: 522415
/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/redis/[connection.py:72](http://connection.py:72/): UserWarning: redis-py works best with hiredis. Please consider installing
  warnings.warn(msg)
[2022-03-17 10:10:16 +0000] [522415] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/[gnupg.py](http://gnupg.py/)", line 881, in __init__
    p = self._open_subprocess(["--version"])
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/[gnupg.py](http://gnupg.py/)", line 961, in _open_subprocess
    result = Popen(cmd, shell=False, stdin=PIPE, stdout=PIPE, stderr=PIPE,
  File "/usr/lib64/python3.8/[subprocess.py](http://subprocess.py/)", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.8/[subprocess.py](http://subprocess.py/)", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'gpg'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/[arbiter.py](http://arbiter.py/)", line 589, in spawn_worker
    worker.init_process()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/workers/[base.py](http://base.py/)", line 134, in init_process
    self.load_wsgi()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/workers/[base.py](http://base.py/)", line 146, in load_wsgi
    self.wsgi = [self.app](http://self.app/).wsgi()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/app/[base.py](http://base.py/)", line 67, in wsgi
    self.callable = self.load()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/app/[wsgiapp.py](http://wsgiapp.py/)", line 58, in load
    return self.load_wsgiapp()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/app/[wsgiapp.py](http://wsgiapp.py/)", line 48, in load_wsgiapp
    return util.import_app([self.app](http://self.app/)_uri)
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/gunicorn/[util.py](http://util.py/)", line 359, in import_app
    mod = importlib.import_module(module)
  File "/usr/lib64/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/pulpcore/app/[wsgi.py](http://wsgi.py/)", line 17, in <module>
    application = get_wsgi_application()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/django/core/[wsgi.py](http://wsgi.py/)", line 12, in get_wsgi_application
    django.setup(set_prefix=False)
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/django/apps/[registry.py](http://registry.py/)", line 114, in populate
    app_config.import_models()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/django/apps/[config.py](http://config.py/)", line 301, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib64/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/pulp_container/app/[models.py](http://models.py/)", line 41, in <module>
    gpg = gnupg.GPG()
  File "/var/lib/pulp/pulpvenv/lib64/python3.8/site-packages/[gnupg.py](http://gnupg.py/)", line 885, in __init__
    raise OSError(msg)
OSError: Unable to run gpg (gpg) - it may not be available.
Worker exiting (pid: 522415)
Shutting down: Master
Reason: Worker failed to boot.
```

**To Reproduce**

**Expected behavior**

**Additional context**
Add any other context about the problem here.  Please provide links to any previous discussions via Discourse or Bugzilla.

### PR DESCRIPTION
**Pull request by ipanova**
closes #640

### CODE REVIEW

File: pulp_container/app/models.py [hunk 1]
```diff
@@ -503,6 +502,7 @@ def validate(self):
             )
 
             with open(signed["signature_path"], "rb") as fp:
+                gpg = gnupg.GPG()
```
Comment (mdellweg) [inline pulp_container/app/models.py:505]: Hmm, this is not a full solution to the problem. Yes it lets the sevices start just to fail later when the actual verification is attempted. But i think it is still a good change.

Comment (ipanova) [reply pulp_container/app/models.py:505]: I don't think we should do anything else.
I think the problem was in the PATH not being set on the user side
```
In [18]: cmd = ['gpg', '--version']

In [19]: result = Popen(cmd, shell=False, env={'PATH': ''})
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-19-f2e3bdace1dd> in <module>
----> 1 result = Popen(cmd, shell=False, env={'PATH': ''})

/usr/lib64/python3.9/subprocess.py in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask)
    949                             encoding=encoding, errors=errors)
    950 
--> 951             self._execute_child(args, executable, preexec_fn, close_fds,
    952                                 pass_fds, cwd, env,
    953                                 startupinfo, creationflags, shell,

/usr/lib64/python3.9/subprocess.py in _execute_child(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, gid, gids, uid, umask, start_new_session)
   1819                     if errno_num != 0:
   1820                         err_msg = os.strerror(errno_num)
-> 1821                     raise child_exception_type(errno_num, err_msg, err_filename)
   1822                 raise child_exception_type(err_msg)
   1823 

FileNotFoundError: [Errno 2] No such file or directory: 'gpg'

In [20]: cmd = ['/bin/gpg', '--version']

In [21]: result = Popen(cmd, shell=False, env={'PATH': ''})

gpg (GnuPG) 2.2.27
libgcrypt 1.9.2-unknown
Copyright (C) 2021 Free Software Foundation, Inc.
License GNU GPL-3.0-or-later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Home: /home/vagrant/.gnupg
Supported algorithms:
Pubkey: RSA, ELG, DSA, ECDH, ECDSA, EDDSA
Cipher: IDEA, 3DES, CAST5, BLOWFISH, AES, AES192, AES256, TWOFISH,
        CAMELLIA128, CAMELLIA192, CAMELLIA256
Hash: SHA1, RIPEMD160, SHA256, SHA384, SHA512, SHA224
Compression: Uncompressed, ZIP, ZLIB, BZIP2
```

Comment (mdellweg) [reply pulp_container/app/models.py:505]: I'm unsure if it's considered good practice to set the PATH environment for daemons though. We should ask an experienced admin. @mikedep333 ?

Comment (ipanova) [reply pulp_container/app/models.py:505]: https://github.com/pulp/pulp_installer/blob/main/roles/pulp_api/templates/pulpcore-api.service.j2#L16
I've asked the user to share the systemd config file for pulpcore-api worker to see what PATH is set there.

Comment (mikedep333) [reply pulp_container/app/models.py:505]: @ipanova Correct, it's invalid syntax. Now the question is whether we (pulp_installer) accidentally put $PATH in it, or whether the user did.

**Review by lubosmj — COMMENTED** [review]
The change does not resolve the problem, as @mdellweg already stated. We are just deferring the actual erroneous behaviour and users will always see that when accessing the `gpg` utility.

How about adding a simple try/except block that does a no-op task with `gpg`? Something like this.
```
try:
    gpg = gnupg.GPG()
    gpg.version
except OSError:
    raise RuntimeError("The path to `gpg` was not configured properly.")

```

**Comment by lubosmj** [comment]
**Seeing an error during a running operation is worse than during the startup.**

**Comment by mdellweg** [comment]
Such a startup check should either live in pulpcore or in every plugin using gpg eventually. The question of having a global check is, whether we declare the dependency on gpg necessary or optional. (If you do not want to anything signed, you can still _use_ Pulp.)

**Comment by ipanova** [comment]
@mdellweg so i've checked and we have the hard dependency declared on gpg in pulpcore https://github.com/pulp/pulpcore/blob/main/requirements.txt#L27 and it's been there at least since pulpcore3.7+

**Comment by mdellweg** [comment]
And in our installation docs we refer to the systemd template that set's the path accordingly. So do you think we can close this as not a bug then?

**Comment by ipanova** [comment]
> And in our installation docs we refer to the systemd template that set's the path accordingly. So do you think we can close this as not a bug then?

I'd still merge this pr so we don't initialize the gpg at the module level

**Comment by mdellweg** [comment]
After a lot of discussion, I'd be fine either way.

**Pull request merged**